### PR TITLE
Fix elevation on windows when the path contains "&" or "'"

### LINKED ISSUE #2809: Elevation issues on windows when the path contains "&" or "'"
- **Etcher version:** 1.5.43
- **Operating system and architecture:** Windows

### PR DESCRIPTION
**Pull request by zvin**
fixes #2809
Change-type: patch

### CODE REVIEW

**Comment by zvin** [comment]
@balena-ci retest

File: tests/shared/permissions.spec.js [hunk 1]
```diff
@@ -32,70 +34,32 @@ describe('Shared: permissions', function () {
...
+            "C:\\Users\\Alice & Bob's Laptop\\\"what\"\\balenaEtcher",
+            [
+              '"a Laser"',
+              'arg1',
+              "'&/ ^ \\",
+              '" $ % *'
+            ],
+            {
+              key: 'value',
+              key2: ' " ^ & = + $ % / \\',
```
Comment (lurch) [inline tests/shared/permissions.spec.js]: Might be also worth adding a `\'` (i.e. an actual single-quote character) to the test string for key2? (both here and on the unix test)

**Comment by lurch** [comment]
Looks like this is changing things to create a temp-script, and then execute that; rather than executing a single in-line command? IIRC there were problems a long time ago with the temp-script approach, which is why it was changed to the in-line command approach?
Perhaps @jviotti remembers more, or perhaps I'm just misremembering things or getting confused with some other issue? :man_shrugging:

File: tests/shared/permissions.spec.js (hunk 1 above)
Comment (zvin) [reply tests/shared/permissions.spec.js]: Added

File: tests/shared/permissions.spec.js [hunk 2]
```diff
@@ -32,70 +34,32 @@ describe('Shared: permissions', function () {
...
+          `set "key=value"${os.EOL}` +
+          `set "key2= " ' ^^ ^& = + $ % / \\"${os.EOL}` +
+          `set "key3=8"${os.EOL}` +
+          `"C:\\Users\\Alice & Bob's Laptop\\\\"what\\"\\balenaEtcher" "\\"a Laser\\"" "arg1" "'&/ ^ \\" "\\" $ % *"`
+        )
       })
     })
 
-    describe('given linux', function () {
+    describe('given unix', function () {
```
Comment (lurch) [inline tests/shared/permissions.spec.js]: You've changed "linux" -> "unix" here, but a couple of lines below it still says `this.osPlatformStub.returns('linux')` ?
I've no idea whether that's important or not though :wink:

**Comment by zvin** [comment]
@balena-ci retest

File: lib/shared/permissions.js [hunk 3]
```diff
@@ -90,68 +97,61 @@ exports.isElevatedUnixSync = () => {
...
+  return `set "${name}=${String(value)}"`
+}
 
-  if (isWindows) {
-    // This is a trick to make the binary afterwards catch
-    // the environment variables set just previously.
-    return _.concat(argv, [ 'call' ])
+// Exported for tests
+exports.createLaunchScript = (command, argv, environment) => {
+  const isWin = os.platform() === 'win32'
```
Comment (lurch) [inline lib/shared/permissions.js]: Might be nice to change `isWin` to `isWindows`, to match code elsewhere?

File: tests/shared/permissions.spec.js (hunk 2 above)
Comment (zvin) [reply tests/shared/permissions.spec.js]: osx and linux should behave the same, updated the tests to test both 'linux' and 'darwin'

File: lib/shared/permissions.js (hunk 3 above)
Comment (zvin) [reply lib/shared/permissions.js]: updated

**Comment by Roboblunder** [comment]
It was not fixed.

**Pull request merged**